geo/src/algorithm/offset/offset_trait.rs

```diff
@@ -0,0 +1,336 @@
+    // Polygon,
+};
+
+/// # Offset Trait
+///
+/// Signed offset of Geometry assuming cartesian coordinate system.
+///
+/// This is a cheap offset algorithm that is suitable for flat coordinate systems
+/// (or if your lat/lon data is near the equator)
+///
+/// My Priority for implementing the trait is as follows:
+/// - [X] Line<impl CoordFloat>
+/// - [X] LineString<impl CoordFloat>
+/// - [X] MultiLineString<impl CoordFloat>
+/// - ... maybe some closed shapes like triangle, polygon?
+///
+/// The following are a list of known limitations,
+/// some may be removed during development,
+/// others are very hard to fix.
+///
+/// - No checking for zero length input.
+///   Invalid results may be caused by division by zero.
+/// - No check is implemented to prevent execution if the specified offset
+///   distance is zero.
+/// - Only local cropping where the output is self-intersecting.
+///   Non-adjacent line segments in the output may be self-intersecting.
+/// - There is no mitre-limit; A LineString which
+///   doubles back on itself will produce an elbow at infinity
+pub trait Offset<T>
+where
+    T: CoordFloat,
+{
+    /// Offset the edges of the geometry by `distance`, where `distance` may be
+    /// negative.
+    ///
+    /// Negative `distance` values will offset the edges of the geometry to the
+    /// left, when facing the direction of increasing coordinate index. For a
+    /// polygon with clockwise winding order, a positive 'offset' corresponds with
+    /// an 'inset'.
+    ///
+    /// ```
+    /// #use crate::{line_string, Coord};
+    /// let input = line_string![
+    ///     Coord { x: 0f64, y: 0f64 },
+    ///     Coord { x: 0f64, y: 2f64 },
+    ///     Coord { x: 2f64, y: 2f64 },
+    /// ];
+    /// let output_expected = line_string![
+    ///     Coord { x: 1f64, y: 0f64 },
+    ///     Coord { x: 1f64, y: 1f64 },
+    ///     Coord { x: 2f64, y: 1f64 },
+    /// ];
+    /// let output_actual = input.offset(1f64);
+    /// assert_eq!(output_actual, output_expected);
+    /// ```
+    fn offset(&self, distance: T) -> Self;
+}
+
+impl<T> Offset<T> for Line<T>
+where
+    T: CoordFloat,
+{
+    fn offset(&self, distance: T) -> Self {
+        let delta = self.delta();
+        let len = (delta.x * delta.x + delta.y * delta.y).sqrt();
+        let delta = Coord {
+            x: delta.y / len,
+            y: -delta.x / len,
+        };
+        Line::new(self.start + delta * distance, self.end + delta * distance)
+    }
+}
+
+
+/// # Offset for LineString
+/// ## Algorithm
+/// Loosely follows the algorithm described by
+/// [Xu-Zheng Liu, Jun-Hai Yong, Guo-Qin Zheng, Jia-Guang Sun. An offset algorithm for polyline curves. Computers in Industry, Elsevier, 2007, 15p. inria-00518005]
+/// (https://hal.inria.fr/inria-00518005/document)
+/// This was the first google result for 'line offset algorithm'
+impl<T> Offset<T> for LineString<T>
+where
+    T: CoordFloat,
+{
+    fn offset(&self, distance: T) -> Self {

```

Comment:
My expectation is that a positive (or outward) offset on a LineString produces a `Polygon` (as per JTS, GEOS, and derived libraries). Could you explain the reasoning here?

Reply:
I have no context on how JTS/GEOS/etc does things, but offsetting a LineString and getting a LineString has many use cases. If you start with something representing the center of a road and want the left or right edge, or the center of a lane, for example. To buffer around a LineString and get a polygon, you could offset inwards (the hole) and outwards (the exterior), right?

Reply:
@dabreegster I'm struggling to visualise that. Do you have an example? The standard implementations of offset (in plain language, **buffer**) operations always produce a polygon from a Line or LineString _as far as I know_. The Shapely docs are representative here: https://shapely.readthedocs.io/en/stable/manual.html#object.buffer

Reply:
Disclaimer that I haven't looked at this implementation, but just a note on terminology...

JTS has an "offset curve" which is a line string to line string transformation. 

> a line lying on one side of another line at a given offset distance

http://lin-ear-th-inking.blogspot.com/2022/01/jts-offset-curves.html?m=1

Reply:
Hi @urschrei thank you so much for all your feedback I really appreciate it, I will do my best to address each one :)

On this @dabreegster has already explained my reasoning really well; To me a buffer operation different to an offset; I am interested in offsetting road center lines, not producing enclosed polygons.

![screenshot](https://github.com/thehappycheese/powerbi-visual-geojson-map-1/raw/main/git_docs/hero_image.jpg)

I recognize that a buffer operation is more generally useful, but I also just need this signed offset function. This offset trait can be extended by a a few clipping and stitching steps to a full buffer operation, this is the end goal of the paper I am following.

Reply:
> I'm struggling to visualise that. Do you have an example?

https://a-b-street.github.io/docs/tech/map/geometry/index.html#projecting-a-polyline, and for the inspiration behind it, https://wwwtyro.net/2019/11/18/instanced-lines.html. The JTS offset curve looks like the same thing. I don't know what this linestring -> linestring operation should be called, but I have plenty of use cases for it, as well as buffering (to get a polygon)

Reply:
Thanks all, it looks like it's just confusion about terminology, mostly on my part. It might be worth discussing whether the current trait name conveys the functionality clearly enough, though maybe it's just a question of ensuring that the module-level docstring does this.

Reply:
@thehappycheese By the way, your screenshot link doesn't appear to work: https://github.com/thehappycheese/powerbi-visual-geojson-map-1/raw/main/git_docs/hero_image.jpg (404)

Reply:
Keen to align terminology with suggestions / other libraries :)

`offset_curve` is nice because it makes it more obvious why this type of signed offset cannot be implemented for a `Point` object. `Point` objects can only have a "buffer" or "outset" applied, not an 'offset_curve'.

Ah apologies for the broken link I picked it from one of my public repos but didn't realize the image link is to a private repo; this one should work?

![Screenshot of road data custom PowerBI visualization](https://github.com/thehappycheese/nicklinref_rust/raw/main/readme_extras/hero_image.jpg)

This is from a custom PowerBI visual I created. I will eventually make that code public when I get a chance to clean up /republish repo. I want the PowerBI visual to dynamically connect to a rust server for a specialized kind of chainage-based  geo-coding.